### ISSUE TITLE
Improve task tap areas

### ISSUE THREAD

**Issue by saraolson** [HabitRPG/habitica-android#1376]
We'd like to extend tap areas on tasks to be a little more generous and capture taps that may have missed the intended UI slightly. We've done an initial extension to make it a little larger but there are some improvements to be made still.

Would like the tap area for a main task button to extend out towards the beginning of the task text, so if you tap anywhere there, it will check off a task.

Also, the UI to open the checklist still behaves a little awkwardly if your tap hits slightly below the button that expands/collapses checklists.

Here's a rough visual of how we'd like the tap areas to act, with the pink areas being tap-able:
<img width="728" alt="Screen Shot 2020-09-29 at 10 59 42 AM" src="https://user-images.githubusercontent.com/12453119/94718852-5122bd00-0320-11eb-8b74-39c817ceedc5.png">

**Comment by Mike-Antonacci** [comment]
Can we add the Accessibility label to this one? As someone with nerve issues in my arms / hands and therefore have fine motor control difficulty sometimes, the lack of good tap areas is making the Android app extremely difficult to use for me. Even when I think I've perfectly aligned my finger with the chat box, probably about 50% of the time it opens the task to edit

**Comment by saraolson** [comment]
@Mike-Antonacci do you know what version you're on? there was a bug that caused task edit to open when tapping the action area in some cases, and it was just fixed in 3.0.1.1

i'll add the accessibility tag to this as well, though.

**Comment by Mike-Antonacci** [comment]
3.0.1 is the version I have, and I don't see an updated version from the Play Store (i.e. I have the most updated one that I can find)

**Comment by saraolson** [comment]
@Mike-Antonacci we started rollout on saturday and just upped it to 100% this morning, so hopefully you will get it within the next 24 hours.

**Comment by Mike-Antonacci** [comment]
I'd like to claim this one.

**Comment by saraolson** [comment]
@Mike-Antonacci its all yours! ill mark it as in progress, let us know if you have any questions.

**Comment by Mike-Antonacci** [comment]
I want to make sure I'm on the right track before I get too in the weeds, and I've found that the hardest part is me getting started. My first hunch as to where I should start is the various layout XML files for each task type, and my second guess is perhaps the drawables for the checklist buttons and whatnot, but I'm thinking that one is not as likely.

Am I on the right track? If yes, I know where to find those layout XML files and can start digging in. If not, if anyone could point me more in the right direction, that would be appreciated

**Comment by phillipthelen** [comment]
the XML files for the task types are daily_item_card, habit_item_card todo_item_card. Currently there already is a method to slightly enlarge the tap area which works through the onTouch method in the BaseTaskRecyclerFragment.

**Comment by Mike-Antonacci** [comment]
Were you referring to the onTouch method found in the BaseTaskViewHolder? I couldn't find anything in TaskRecyclerViewFragment but did find a few uses of dpToPx that (at first glace) looks like it does what you're describing. I'll play around with it.

**Comment by Mike-Antonacci** [comment]
<img width="728" alt="current behavior" src="https://user-images.githubusercontent.com/70340774/100464320-a06c3c00-308a-11eb-94b8-022dc10264e4.png">

@saraolson here's what I have so far, let me know if this is considered acceptable behavior or not (I'm thinking not, but want to be sure before I make more changes)

The green rectangle on the left represents the clickable area for the top level checkbox, meaning that if you click to the right of any of the items on the checklist, it instead checks off the entire Daily. I'm assuming that is not acceptable. secondly, and this is not reflected in the screenshot, the clickable area on the left goes all the way as far as the entire first letter ("D" in "Do"); is that too far?

The orange checkbox on the right represents clickable area that collapses the entire checklist. Is that acceptable? Or do we want to trim that one up?

**Comment by Mike-Antonacci** [comment]
@saraolson @phillipthelen I'm getting unexpected behavior: Even without changing any code, the area in orange on the right collapses the checklist and the area in green on the left checks off the overall task (not any of the subtasks, even when next to them). In contrast, my standard Habitica app (not the debug app) does neither: the orange area does nothing, and the green area checks off the appropriate subtask if you're next to it. In essence, extra task tap area is getting added without me making any changes.
<img width="728" alt="current behavior" src="https://user-images.githubusercontent.com/70340774/100489426-50649800-30d1-11eb-97ca-2a77d40732e1.png">

This happens on three different virtual devices and and on my physical device when connected to Android Studio. Any suggestions on how I should proceed? Testing is difficult since it's hard to distinguish between what is my change and what is happening spontaneously.

**Comment by Mike-Antonacci** [comment]
Okay, so I think I figured out why the above behavior is happening, but I don't know what to do about it. I put in some Log messages to the onTouch method, which spit out the X and Y coordinates of a motion event, the height the code is looking for (for y-coordinate), and the width the code is looking for (for x-coordinate).

What I discovered is that when a checklist is expanded, the height of the mainTaskWrapper also increases, and so the code is looking for a Y coordinate that is within the overall height of the fully expanded checklist, and therefore registers the touch as being for the highest level of the checklist (i.e. for left touch, the upper left item, and for right touch, collapsing the checklist) rather than for the checklist item right next to it.

So, if grabbing the height of the mainTaskWrapper won't lead to the desired behavior if the checklist is expanded, is there another height I could grab instead?

Where I could use some input is how the parent/child relationship works within checklist views (how is the checklist itself "built"? Since the overall task contains checklist items, how does that containing happen? I know that once we are in the ChecklistedViewHolder, `onLeftActionTouched()` gets overridden to apply to the checklist item, but I don't know how the ChecklistedViewHolder fits in with the mainTaskWrapper )

**Comment by saraolson** [comment]
yea its definitely not intended that the main task tap areas would trail down the checklist. i can't offer much technical information about how checklists are built, @phillipthelen will have to speak on that, but here's the design spec i made up when we redid checklists originally, if it helps: 
<img width="782" alt="Screen Shot 2020-12-01 at 3 19 02 PM" src="https://user-images.githubusercontent.com/12453119/100792406-9b8bed00-33e8-11eb-985b-c30454ff5253.png">

so, when a task is expanded, this is basically tacked on to the bottom edge and the rounded corner moves down with it.

**Comment by saraolson** [comment]
@Mike-Antonacci 
1. yep the yellow orange 8 is always 8
2. yup 8dp
3. clickable area for the first checklist item
4. yep 40 is minimum. the way the row expands right now for multiple lines works well, so wouldnt want to change that.
5. yeah rounded corner on bottom left is clickable for last checklist item

**Comment by Mike-Antonacci** [comment]
@saraolson no matter how many ways I try, I'm unable to get it to work 100% for checklists. The best I can do is expand the task tap area for the main task of a checklist but do nothing for the checklist that items (any expansion I do on their tap area instead expands the main task tap area). I even have Log messages that detect the X and Y coordinates of each tap and then output a message saying whether it's the main or the checklist areas, the messages work correctly (i.e. when I click in the checklist area, the message says I clicked in a checklist area) but even still, it checks off the main task instead of the checklist.

Two options for you:
1) I can submit what I have is a partial fix and let someone else build on what I've done
2) I can scrap it completely and let whoever claims this next work from scratch and not have to use what I did

I'm fine either way, just let me know if I should do a pull request for what I've done so far.

**Comment by saraolson** [comment]
@Mike-Antonacci i'd be interested in at least trying out what you've worked on to see how it feels in the app. I'd go ahead and submit it and Phillip and I can figure out how we'd like to handle it from there :) Thanks for all the work you put into this tricky issue. We tossed it to github because it had given us so much trouble to begin with lol

**Comment by Mike-Antonacci** [comment]
Screencast demonstrating the changes

https://user-images.githubusercontent.com/70340774/103835815-7a03ce00-5044-11eb-8fcb-091b64d3db50.mp4

**Comment by Mike-Antonacci** [comment]
also, here is the pull request: https://github.com/HabitRPG/habitica-android/pull/1454

The version of the code for the pull request is the one in the video above, where the if statements or what I think *should* work but are not. If you'd rather me do a pull request (or a separate pull request) for the version where I purposely don't expand the task tap area for checklist items (only the main task item), I can do that as well.

**Comment by Mike-Antonacci** [comment]
@phillipthelen @saraolson 

I noticed you merged pull request for my partial fix for this issue. The version I submitted will cause issues for checklist items, they will instead check off the main task item. (I submitted this version and not another because the logic is most clear and would be most useful to someone else)

I recommend that someone comment out lines 268-272 of BaseTaskViewHolder before this rolls out to the public, which will result in checklist items not having an expanded tap area at all. I'd be happy to do it if you want.

From there, a future developer can build on what I've done and see if they can solve expanding the tap area for checklist items

**Comment by phillipthelen** [comment]
Sorry. I misunderstood your comment. I'll take another look at the code and comment out the parts that don't work properly and see if I can finish this up to work better.
